File: frontend/bcaiaccount/lib/model/attachment_model.dart
import 'package:json_annotation/json_annotation.dart';

part 'attachment_model.g.dart';

@JsonSerializable()
class AttachmentModel {
  @JsonKey(name: 'id')
  String? id;

  @JsonKey(name: 'shopid')
  String shopid;

  @JsonKey(name: 'screen_type')
  String screenType;

  @JsonKey(name: 'docno')
  String docno;

  @JsonKey(name: 'guidfixed')
  String guidfixed;

  @JsonKey(name: 'filename')
  String filename;

  @JsonKey(name: 'original_name')
  String originalName;

  @JsonKey(name: 'content_type')
  String contentType;

  @JsonKey(name: 'file_type')
  String fileType;

  @JsonKey(name: 'size')
  int size;

  @JsonKey(name: 'description')
  String? description;

  @JsonKey(name: 'uploaded_by')
  String uploadedBy;

  @JsonKey(name: 'uploaded_name')
  String? uploadedName;

  @JsonKey(name: 'created_at')
  String createdAt;

  @JsonKey(name: 'updated_at')
  String updatedAt;

  @JsonKey(name: 'url')
  String? url; // Presigned URL

  AttachmentModel({
    this.id,
    required this.shopid,
    required this.screenType,
    required this.docno,
    required this.guidfixed,
    required this.filename,
    required this.originalName,
    required this.contentType,
    required this.fileType,
    required this.size,
    this.description,
    required this.uploadedBy,
    this.uploadedName,
    required this.createdAt,
    required this.updatedAt,
    this.url,
  });

  factory AttachmentModel.fromJson(Map<String, dynamic> json) => _$AttachmentModelFromJson(json);

  Map<String, dynamic> toJson() => _$AttachmentModelToJson(this);
}

@JsonSerializable()
class AttachmentListResponse {
  @JsonKey(name: 'status')
  String status;

  @JsonKey(name: 'code')
  int code;

  @JsonKey(name: 'count')
  int count;

  @JsonKey(name: 'total')
  int total;

  @JsonKey(name: 'data', defaultValue: [])
  List<AttachmentModel> data;

  AttachmentListResponse({
    required this.status,
    required this.code,
    required this.count,
    required this.total,
    this.data = const [],
  });

  factory AttachmentListResponse.fromJson(Map<String, dynamic> json) => _$AttachmentListResponseFromJson(json);

  Map<String, dynamic> toJson() => _$AttachmentListResponseToJson(this);
}

@JsonSerializable()
class AttachmentUploadResponse {
  @JsonKey(name: 'status')
  String status;

  @JsonKey(name: 'code')
  int code;

  @JsonKey(name: 'message')
  String? message;

  @JsonKey(name: 'data')
  AttachmentModel? data;

  AttachmentUploadResponse({
    required this.status,
    required this.code,
    this.message,
    this.data,
  });

  factory AttachmentUploadResponse.fromJson(Map<String, dynamic> json) => _$AttachmentUploadResponseFromJson(json);

  Map<String, dynamic> toJson() => _$AttachmentUploadResponseToJson(this);
}

File: backend/internal/goapi/models/attachment-model.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// AttachmentMetadata - ข้อมูลไฟล์แนบเอกสาร (PO, Sale, Purchase, etc.)
type AttachmentMetadata struct {
	ID           primitive.ObjectID `bson:"id,omitempty" json:"id"`
	HoldingCode  string             `bson:"holdingcode" json:"holdingcode"`
	ScreenType   string             `bson:"screentype" json:"screentype"`                       // purchaseorder, sale, purchase, etc.
	DocNo        string             `bson:"docno" json:"docno"`                                 // เลขที่เอกสาร
	GuidFixed    string             `bson:"guidfixed" json:"guidfixed"`                         // GUID ของเอกสาร
	FileName     string             `bson:"filename" json:"filename"`                           // ชื่อไฟล์ใน R2 (timestamp_hash + extension)
	OriginalName string             `bson:"originalname" json:"originalname"`                   // ชื่อไฟล์เดิม
	ContentType  string             `bson:"contenttype" json:"contenttype"`                     // MIME type
	FileType     string             `bson:"filetype" json:"filetype"`                           // pdf, xlsx, jpg, png
	Size         int64              `bson:"size" json:"size"`                                   // ขนาดไฟล์ (bytes)
	R2Key        string             `bson:"r2key" json:"-"`                                     // key ใน R2 bucket (ไม่ส่งออก)
	Description  string             `bson:"description,omitempty" json:"description,omitempty"` // คำอธิบาย
	UploadedBy   string             `bson:"uploadedby" json:"uploadedby"`                       // usercode
	UploadedName string             `bson:"uploadedname" json:"uploadedname"`                   // user_name
	CreatedAt    time.Time          `bson:"createdat" json:"createdat"`
	UpdatedAt    time.Time          `bson:"updatedat" json:"updatedat"`
}

// AttachmentUploadRequest - request สำหรับ upload attachment
// Form data: file, holdingcode, screentype, docno, guidfixed, description, uploadedby, uploadedname
type AttachmentUploadRequest struct {
	File         string `form:"file"` // multipart file
	HoldingCode  string `form:"holdingcode"`
	ScreenType   string `form:"screentype"`
	DocNo        string `form:"docno"`
	GuidFixed    string `form:"guidfixed"`
	Description  string `form:"description"`
	UploadedBy   string `form:"uploadedby"`
	UploadedName string `form:"uploadedname"`
}

// AttachmentListRequest - request สำหรับ list attachments
type AttachmentListRequest struct {
	HoldingCode string `json:"holdingcode"`
	ScreenType  string `json:"screentype,omitempty"`
	DocNo       string `json:"docno,omitempty"`
	GuidFixed   string `json:"guidfixed,omitempty"`
	Limit       int64  `json:"limit,omitempty"`
	Skip        int64  `json:"skip,omitempty"`
}

// AttachmentDeleteRequest - request สำหรับ delete attachment
type AttachmentDeleteRequest struct {
	HoldingCode  string `json:"holdingcode"`
	AttachmentID string `json:"attachmentid,omitempty"`
	FileName     string `json:"filename,omitempty"`
}

// AttachmentResponse - response structure
type AttachmentResponse struct {
	Status  string              `json:"status"`
	Code    int                 `json:"code"`
	Message string              `json:"message,omitempty"`
	Data    *AttachmentMetadata `json:"data,omitempty"`
}

// AttachmentListDataItem - attachment item with private backend URL
type AttachmentListDataItem struct {
	ID           primitive.ObjectID `json:"id"`
	HoldingCode  string             `json:"holdingcode"`
	ScreenType   string             `json:"screentype"`
	DocNo        string             `json:"docno"`
	GuidFixed    string             `json:"guidfixed"`
	FileName     string             `json:"filename"`
	OriginalName string             `json:"originalname"`
	ContentType  string             `json:"contenttype"`
	FileType     string             `json:"filetype"`
	Size         int64              `json:"size"`
	Description  string             `json:"description,omitempty"`
	UploadedBy   string             `json:"uploadedby"`
	UploadedName string             `json:"uploadedname"`
	CreatedAt    time.Time          `json:"createdat"`
	UpdatedAt    time.Time          `json:"updatedat"`
	URL          string             `json:"url,omitempty"` // Private backend URL
}

// AttachmentListDataResponse - response สำหรับ list attachments
type AttachmentListDataResponse struct {
	Status string                   `json:"status"`
	Code   int                      `json:"code"`
	Count  int                      `json:"count"`
	Total  int64                    `json:"total"`
	Data   []AttachmentListDataItem `json:"data"`
}
